Add support for Recycle Bin

gave92 opened the pull request:
Hello this pull request adds support for browsing the **Recycle Bin**!
"Support" here means that the recycle bin can be browsed just like any other folder in Files UWP, it's not just a "link" that opens the recycle bin in explorer.

![image](https://user-images.githubusercontent.com/9673091/82843494-91d77900-9edd-11ea-8699-1d43d4045238.png)

#### How it works
As I'm not aware of any way to properly access the recycle bin folder from the UWP app, I'm making use of Files UWP full trust process. The fulltrust process enumerates the contents of the recycle bin using the Shell32 api and send the results to the UWP app. Communication between the fulltrust process and the UWP app is done through AppService.

*NB: This requires to change the fulltrust process lifecyle*: instead of starting a new instance every time to handle the commands to open CMD, toggle QuickLook, etc., a single instance keeps running until the UWP app is closed or suspended.

#### What works
- New entry on the sidebar with localized name of the recycle bin
- Enumerating recycle bin elements
- Monitor filesystem changes and refresh items
- Showing the thumbnails for deleted files
- Properties dialog for deleted files
- Deleting files from the recycle bin

#### What's missing
- A button to empty recycle bin
- A customized context menu with the proper options (e.g "restore")

Let me know what you think of this, if you approve of the "architectural" changes I can keep working to complete what's missing.

duke7553 commented:
@gave92 Have you tried accessing the recycling bin by it's GUID with the existing method? I'd assume it works since we currently use the Win32 FindFirstFileFromApp in the Filesystem view model.

Recycle Bin GUID - `::{645FF040-5081-101B-9F08-00AA002F954E}`

yaichenbaum commented:
Support for recycle bin would be great, I am not sure if we want to keep it pinned by default. Can you add a feature flag that turns on/off this feature?

gave92 commented:
@yaichenbaum yup I can do that, perhaps also add it to the New Tab page?

@duke7553 I've tried using the Shell32 api from the UWP side (meaning using the functions [SHGetFolderLocation](https://docs.microsoft.com/it-it/windows/win32/api/shlobj_core/nf-shlobj_core-shgetfolderlocation) and other [IShellFolder](https://docs.microsoft.com/en-us/windows/win32/api/shobjidl_core/nn-shobjidl_core-ishellfolder) related methods and for some reasons I cannot enumerate the contents of the recycle bin. The same method works when executed in the fulltrust process.
Is this what you were suggesting by "accessing the recycling bin by it's GUID"?

yaichenbaum commented:
@gave92 I would add it to the new page, but it should be behind the feature flag as well.

gave92 commented:
@yaichenbaum I'll do that. I'm gonna wait a bit though, until I get some feedback on proposed method of accessing the recycle bin (fulltrust process + appservice).
Hopefully someone can think of a simpler way.

tsvietOK commented:
@gave92 
First
Repro steps:
1. Open recycle bin
2. Navigate to folder inside it
3. Press Up button on navigation panel
4. Permission dialog displayed 
Second
I think folders in recycle bin should not have ability to be opened, renamed, pinned to sidebar, shared, opened in new windows/tab.

tsvietOK commented:
And last, i don't think it is a good idea to merge this PR without supporting restore and clear recycle bin functions.

tsvietOK commented:
Other way, mark this feature as Experimental

yaichenbaum commented:
> And last, i don't think it is a good idea to merge this PR without supporting restore and clear recycle bin functions.

We can do this behind a feature flag, this way we are not making any commitment to the feature and the capabilities can be changed.

ADeltaX commented on Files/Program.cs:
```diff
@@ -43,10 +41,14 @@ static void Main()
-            ApplicationData.Current.LocalSettings.Values["Arguments"] = "ShellCommand";
-            ApplicationData.Current.LocalSettings.Values["pid"] = pid;
-            await FullTrustProcessLauncher.LaunchFullTrustProcessForCurrentAppAsync();
+            if (App.Connection != null)
```
App.Connection here is always null because it checks immediately at launch (hence the app will not work as shell replacement).

gave92 replied:
Thanks, I reverted the OpenShellCommandInExplorer function to the original. There's no need to set up the AppService stuff when we're just going to kill the process

ADeltaX replied:
The `return` is sufficient only for Debug mode. In release mode this does not close the window, hence why we need to kill the "zombie" process

gave92 commented:
Thanks for the feedbacks.

@tsvietOK
1. Permission dialog displayed => will be fixed when disallowing navigation inside recycle bin

2. I think folders in recycle bin should not have ability to be opened, renamed, pinned to sidebar, shared, opened in new windows/tab => agreed, need to customize the context menu and buttons (e.g add an "Empty recycle bin" button and a "Restore" to context menu)

3. "Recycle Bin" tab text untranslated => solved

4. Button "Copy path" causes app crash => I cannot reproduce this, but it will go away when we change the buttons/context menu (see 2)

tsvietOK commented on Files.Launcher/Program.cs:
```diff
@@ -1,85 +1,319 @@
+                            var file_list = new List<ShellFileItem>();
+                            for (int i = 0; i < items.Count; i++)
+                            {
+                                Shell32.FolderItem FI = items.Item(i);
```
If this vars are inside loop, so how to use camelCase?

tsvietOK replied:
You forget to change this virus names to camelCase. (Currently they are used CamelCase)

tsvietOK commented on Files/Views/InstanceTabsView.xaml.cs:
```diff
@@ -131,6 +134,11 @@ public async void AddNewTab(Type t, string path)
                 }
+                else if (path.StartsWith(App.AppSettings.RecycleBinPath, StringComparison.OrdinalIgnoreCase))
+                {
+                    tabLocationHeader = ApplicationData.Current.LocalSettings.Values.Get("RecycleBin_Title", "Recycle Bin");
```
What about to use localSettings.Values?

yaichenbaum replied:
Shouldn't we use `App.AppSettings`?

gave92 replied:
Perhaps it would be better just to add the localized recycle bin name to the translation files?

tsvietOK commented on Common/ShellFileItem.cs:
```diff
@@ -0,0 +1,38 @@
+{
+    public class ShellFileItem
+    {
+        public bool isFolder;
```
UpperCamelCase for public fields

gave92 replied:
Ok fixed the naming convention of a few variables. I'll be more careful 😅

yaichenbaum commented:
- By default, recycle bin should not be pinned to the sidebar.
- The app crashes when navigating to recycle bin.

gave92 commented:
@yaichenbaum Ok I'll change the default to not pinned.
I'm aware of the crash, I'm waiting for the owner of the [Vanara](https://github.com/dahall/Vanara) I'm now referencing to publicly release the bug fixed version.

yaichenbaum commented:
@gave92 Is there a way I can test this out in the meantime?

gave92 commented:
@yaichenbaum Yes please try again with the latest commit.
Notes:
* auto refresh of recycle bin contents is disabled until the release of the updated Vanara library
* creating/pasting/... a file in the recycle bin crashes the app, I'm not fixing this as the solution will be to have a different context menu and buttons for the recycle bin

yaichenbaum commented:
@gave92 I can now open recycle bin but it appears to be empty.

gave92 commented:
@yaichenbaum Can you check whether while you are in the recycle bin folder (in Files UWP) there's a process called `FilesFullTrust.exe` running in Task Manager?

yaichenbaum commented:
![image](https://user-images.githubusercontent.com/39923744/83160993-dc234a80-a0d5-11ea-81e6-77c5206cbfe3.png)

yaichenbaum commented:
@gave92 I do not see `FilesFullTrust.exe` in Task Manager.

gave92 commented:
Thanks for the feedback, that means the the process has crashed so you see no results on the UWP app. Please confirm whether the process is there when you start the app (before navigating to the reycle bin). Anyway I'll add some logging to understand the issue.

gave92 commented:
Thanks for the feedback, the issue should now be fixed!

duke7553 commented on Files/View Models/ItemViewModel.cs:
```diff
@@ -274,7 +277,8 @@ public ItemViewModel()
             //(App.CurrentInstance as ProHome).RibbonArea.RibbonViewModel.AlwaysPresentCommands.PropertyChanged += AlwaysPresentCommands_PropertyChanged;
-            _cancellationTokenSource = new CancellationTokenSource();
+            _addFilesCTS = new CancellationTokenSource();
+            _semaphoreCTS = new CancellationTokenSource();
```
semaphore?

gave92 commented:
@duke7553 The feature *is* behind a flag, but anyone can enable it in the settings. So perhaps it should be put under the Experimental setting page? And made disabled like the "Show hidden files and folders" setting?
Note: even if the recycle bin sidebar folder is not shown the modifications to the fulltrust process remain in effect. Is this acceptable? Are you planning to merge this is its current state?

duke7553 commented:
@gave92 Thanks for clarifying this. As long as the recycle bin location is only visible when a flag inside the Experimental Settings page is toggled from its default state, we can merge this when you've finished work on it.

gave92 commented:
A little update on this:
* Added customized context menu with "Empty recycle bin" action
* Added customized context menu for recycle bin items with "Restore" action
* Disabled menu bar "New file" action
* Disabled renaming of items in recycle bin
* Disabled navigation and opening of items in recycle bin
* The option to enable recycle bin support is now in the Experimental settings page

Let me know what you think!

yaichenbaum commented:
@duke7553 @gave92 I should have clarified the first time that when I said feature flag, I meant the experimental settings page.

yaichenbaum commented:
@gave92 The app doesn't crash but it still won't load any items I have in recycle bin.

gave92 commented:
Curious. Please check the following:
* That no "FilesFullTrust.exe" process is running after you visit the bin from the uwp app
* In the application data path ("%localappdata%\packages\FilesUWPxxxxx\LocalState" there should be a "debug_fulltrust.txt" with the logged exception

yaichenbaum commented:
@gave92 FilesFullTrust.exe is running, I see the regular debug file but nothing for debug_fulltrust.exe.

gave92 commented:
Ok so the fulltrust process is not crashing like before which is good... and now I'm confused.
* Does opening files from the uwp work? Does the command "Open in terminal" work?
* Would you be willing to run the app from this PR in debug mode in visual studio?

yaichenbaum commented:
I don't get a crash, it's just that the progress bar keeps loading in recycle bin and doesn't load any items that are there.

yaichenbaum commented:
@gave92 It finally loaded after a few minutes, I think we need to figure out a faster way to load recycle bin before shipping it.

gave92 commented:
Ok just how much stuff is in your recycle bin? xD

I'll test with a lot of files as well. Thanks for finding out the issue.

yaichenbaum commented:
@gave92 It is much faster now.

yaichenbaum commented on Common/ShellFileItem.cs:
```diff
@@ -0,0 +1,38 @@
+        public ulong FileSizeBytes;
+        public string FileType;
+
+        public ShellFileItem()
```
Is this necessary?

yaichenbaum commented on Common/ShellFileItem.cs:
```diff
@@ -0,0 +1,38 @@
+        public ulong FileSizeBytes;
+        public string FileType;
+
+        public ShellFileItem()
+        {
+
+        }
```
Is this necessary?

gave92 replied:
I think so. I'm using Json.Net to serialize/deserialize ShellFileItem objects and as far as I know Json.Net need a parameterless constructor to to this.

yaichenbaum commented on Files/BaseLayout.cs:
```diff
@@ -189,11 +192,44 @@ private void UnloadMenuFlyoutItemByName(string nameToUnload)
+            {
+                (this.FindName("EmptyRecycleBin") as MenuFlyoutItemBase).Visibility = Visibility.Visible;
+                (this.FindName("OpenTerminal") as MenuFlyoutItemBase).IsEnabled = false;
+                UnloadMenuFlyoutItemByName("NewEmptySpace");
```
You will probably need to use the localized strings to find the right items here.

gave92 replied:
I'm finding elements based on the "x:Name" attribute. That does not change based on the locale.

yaichenbaum commented on Files/BaseLayout.cs:
```diff
@@ -256,6 +292,28 @@ public void RightClickContextMenu_Opening(object sender, object e)
+            UnloadMenuFlyoutItemByName("CopyItem");
+            UnloadMenuFlyoutItemByName("RenameItem");
+            UnloadMenuFlyoutItemByName("SidebarPinItem");
+            UnloadMenuFlyoutItemByName("QuickLook");
```
Being that we are working on adding support for more menu items, it might be a better idea to unload everything through a loop rather than removing the known items, this way we can add more items without updating this list.

gave92 replied:
You mean first removing everything and than adding just what's needed for the recycle bin? Yup I can do that

yaichenbaum replied:
Correct, instead of removing specific ones, remove everything and then add the one we need.

gave92 replied:
I did it a little differently, let me know what you think: I simply created a specialized context menu for recycle bin items so there is no need to Hide/Show menu items and there is no issue when more menu items are added to the "standard" menu.

yaichenbaum replied:
Doing it as an individual context menu is a great idea.
